src/note/nullifier.rs

```diff
@@ -0,0 +1,30 @@
+use halo2::arithmetic::CurveExt;
+use pasta_curves::pallas;
+
+use super::NoteCommitment;
+use crate::{
+    keys::NullifierDerivingKey,
+    spec::{extract_p, mod_r_p},
+};
+
+/// A unique nullifier for a note.
+#[derive(Debug)]
+pub struct Nullifier(pub(super) pallas::Base);
+
+impl Nullifier {
+    /// $DeriveNullifier$.
+    ///
+    /// Defined in [Zcash Protocol Spec § 4.16: Note Commitments and Nullifiers][commitmentsandnullifiers].
+    ///
+    /// [commitmentsandnullifiers]: https://zips.z.cash/protocol/nu5.pdf#commitmentsandnullifiers
+    pub(super) fn derive(
+        rho: pallas::Base,
+        psi: pallas::Base,
+        cm: NoteCommitment,
+        nk: &NullifierDerivingKey,
+    ) -> Self {
+        let k = pallas::Point::hash_to_curve("z.cash:Orchard")(b"K");
```

Comment:
We should hard-code this point and check that it's correct in a test, but this is fine for now.